### DISCUSSION TITLE
How to setup worker.php

### DISCUSSION THREAD

**Discussion by mokraemer** [tine20/tine20#7369] [Q&A] [answered]
In the installation guide, worker.php was not mentioned. As far as I know this is needed to invite people to appointments.

What interval should the worker triggered? Should it be called directly via cron or via http?
btw. it has some addtional requirements - php-posix needs to be activated.

Any guides?

Since this module was not active for some time, how do I clear the queue, so I don't want to send mails for the last 5 years....

**Comment by lab-at-nohl** [comment, marked as answer]
First, given the fact that no mails are sent currently*) I guess you have enabled redis as asynchronous queue. Thus, you have to clear redis (the queue) indeed. I will have a look later, if an ordinary shell call is required or some Tine command exists.**) 

*) For test purposes you can try to disable redis and set external participants. Than you can see the problem solved by the queue, your script will finish after all mails are sent. 

**) EDIT: Try `redis-cli flushdb` to delete **all** pending records. However, you might want to do `redis-cli keys '*' | grep -v CACHE` before to make sure nothing important is flushed. I remember a discussion about a cli command to do it, but I cannot find the outcome. 

Second, if you are on Linux worker.php is a Daemon that has to be controlled by systemd (systemctl). You may find some hints on his to create such unit here:
https://github.com/tine20/tine20/issues/6929
Eventually, you do not have to trigger. Systemd does it and even respawn if the process dies. 

If you are not root or privileged to add unit files, we could try to figure out a bash loop and lock files; I used such a setup but do not recommend it.

Third, besides the queue you still need a Cron job for asynchronous jobs. Having two systems is very unfortunate but that is how it is.
Edit 2: Your cron job must include a line like `*/2 * * * * /usr/bin/php -d memory_limit=512M /path/to/tine20.php --method Tinebase.triggerAsyncEvents --username cronuser` to invoke those jobs. 

Please look at the linked issue. If any questions remain I am happy to help.

Johannes

**Reply by mokraemer** [reply]
Thanks Joannes,
I was following the setup and the guide, but it was not said, what is cached in redis. Usally caching helps for performance, so I activated it. Even [here](https://github.com/tine20/tine20/wiki/EN:Install-Howto) it is not explained you have to setup a cron or the worker in cases you use it. I don't want to complain - I just think it has to be added. In the current situation it looks like a bug, since no mails are send "by default".

**Reply by mokraemer** [reply]
Is it right, if I disable redis, I will not need the worker? Besides sessions, what I was aware of, is there anything else that is stored there, and has performance impacts?

**Reply by lab-at-nohl** [reply]
> Is it right, if I disable redis, I will not need the worker?

To be honest, I don't know for sure. Redis is used for ActionQueue, Caching and Session files. The latter two can be set to use files as backend instead; but I don't know what happens if you disable ActionQueue completely. In my understanding all possible actions would be handled synchronous in the main thread. Thus, it may slow the response down but would not break anything. Also you do not need worker.php anymore. 

I did a quick test on my production machine (create event with external attendee):

- Disabled ActionQueue in config.inc.php --> mails to external are sent and received
- Stopped worker.php without disabling ActionQueue --> no mails are sent until restart of worker.php

Hence, sending invites to externals works without ActionQueue. 

Besides ActionQueue we have AsyncEvents which are to be triggered by Cron. I use both without questioning it since years. However, reading the code Cron can be disabled as well. But all that does is to skip the execution of triggerAsncEvents. Thus, it does not look save.
https://github.com/tine20/tine20/commit/7de4525b6b31a4ad6e701900785fe3e4f4a716f6

While worker.php was disabled, triggerAsyncEvents did not release the invites. 

> no mails are send "by default".

Given my testing I would expect your setup problem not with the queue. You should be able to send invites to externals even though ActionQueue is disabled...

**Reply by lab-at-nohl** [reply]
And one little note: The "Tine 2.0 notification service" does not use your personal mail address. It uses the address configured in setup.php for notifications! Maybe you should have a look there.

**Comment by mokraemer** [comment]
Thanks a lot. I'll test it today and see if it works :)

**Comment by mokraemer** [comment]
it "works" without the actionqueue. The mails are sent from the account itself, which is much better than a general mailadress.
But if I accept/decline an appointment, I have a mail in my inbox, but tine does not update the event. (the cron-job is running - if it should be triggered by it)
